File: backend/app/core/audit_store.py
"""In-memory hash-chained audit store — port of src/lib/audit-store.ts.

SHA-256 chain (prevHash + canonical payload) with the same four seed entries.
"""
import hashlib
import json
import random
import string
import time
from datetime import datetime, timedelta, timezone

GENESIS_HASH = "0" * 64
MAX_ENTRIES = 500

_store: list[dict] = []
_head_hash = GENESIS_HASH
_chain_counter = 0


def _now_iso(offset_hours: float = 0) -> str:
    dt = datetime.now(timezone.utc) + timedelta(hours=offset_hours)
    return dt.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


def _compute_chain_hash(prev_hash: str, entry: dict) -> str:
    canonical = json.dumps(
        {
            "id": entry["id"],
            "timestamp": entry["timestamp"],
            "user": entry["user"],
            "email": entry["email"],
            "action": entry["action"],
            "module": entry["module"],
            "details": entry["details"],
            "type": entry["type"],
        },
        separators=(",", ":"),
    )
    return hashlib.sha256((prev_hash + canonical).encode()).hexdigest()


def add_audit_entry(*, user: str, email: str, action: str, module: str, details: str, type: str) -> dict:
    global _head_hash, _chain_counter
    suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=4))
    entry = {
        "id": f"evt-{int(time.time() * 1000)}-{suffix}",
        "timestamp": _now_iso(),
        "user": user,
        "email": email,
        "action": action,
        "module": module,
        "details": details,
        "type": type,
    }
    entry["prevHash"] = _head_hash
    entry["chainHash"] = _compute_chain_hash(_head_hash, entry)
    entry["chainIndex"] = _chain_counter
    _head_hash = entry["chainHash"]
    _chain_counter += 1
    _store.insert(0, entry)  # newest first for display
    del _store[MAX_ENTRIES:]
    return entry


def get_audit_entries(limit: int = 50) -> list[dict]:
    return _store[:limit]


def get_audit_count() -> int:
    return len(_store)


def verify_audit_chain() -> dict:
    ordered = sorted(_store, key=lambda e: e["chainIndex"])
    running = GENESIS_HASH
    for entry in ordered:
        if entry["prevHash"] != running or entry["chainHash"] != _compute_chain_hash(running, entry):
            return {
                "valid": False,
                "totalEntries": len(_store),
                "verifiedAt": _now_iso(),
                "breakAt": entry["id"],
                "breakIndex": entry["chainIndex"],
            }
        running = entry["chainHash"]
    return {"valid": True, "totalEntries": len(_store), "verifiedAt": _now_iso()}


def _seed() -> None:
    seeds = [
        ("Marcus Webb", "m.webb@ligaturerd.io", "User Login", "Auth",
         "Authenticated via SSO — session opened", "login", -8),
        ("Sarah Chen", "sarah.chen@ligaturerd.io", "Document Approved", "Submissions",
         "Approved CSR for LIG-2847 — three-stage review complete", "document", -6),
        ("James Liu", "j.liu@ligaturerd.io", "Status Changed", "Safety",
         "ICSR-2025-0088 status → Submitted (transmitted to EMA gateway)", "document", -4),
        ("Priya Anand", "p.anand@ligaturerd.io", "CAPA Created", "QMS",
         "CAPA-2026-0042 opened — deviation source: batch documentation gap", "document", -2),
    ]
    global _head_hash, _chain_counter
    for index, (user, email, action, module, details, type_, hours) in enumerate(seeds):
        entry = {
            "id": f"seed-{index + 1}",
            "timestamp": _now_iso(hours),
            "user": user,
            "email": email,
            "action": action,
            "module": module,
            "details": details,
            "type": type_,
        }
        entry["prevHash"] = _head_hash
        entry["chainHash"] = _compute_chain_hash(_head_hash, entry)
        entry["chainIndex"] = index
        _head_hash = entry["chainHash"]
        _chain_counter = index + 1
        _store.insert(0, entry)


_seed()

File: frontend/src/lib/audit-store.ts

// =============================================================================
// SERVER-SIDE AUDIT LOG STORE — v0.98.0
// =============================================================================
// SHA-256 hash chain: every entry links to its predecessor via chainHash.
// chainHash = SHA-256( prevHash + canonicalized entry payload )
// This makes the log tamper-evident: any edit breaks all subsequent hashes.
// =============================================================================

import crypto from 'crypto';

export interface AuditEntry {
  id: string;
  timestamp: string; // ISO string
  user: string;
  email: string;
  action: string;
  module: string;
  details: string;
  type: 'login' | 'invite' | 'accept' | 'permission' | 'system' | 'document';
  // Chain integrity fields (Part 11 §11.10(e))
  prevHash: string;    // chainHash of the previous entry (GENESIS_HASH for first)
  chainHash: string;   // SHA-256( prevHash + canonical payload )
  chainIndex: number;  // 0-based position in insertion order
}

export interface ChainVerifyResult {
  valid: boolean;
  totalEntries: number;
  verifiedAt: string;
  breakAt?: string; // id of the first entry that fails verification
  breakIndex?: number;
}

// The genesis constant — prepended to the very first entry's hash input
const GENESIS_HASH = '0000000000000000000000000000000000000000000000000000000000000000';

const MAX_ENTRIES = 500;

// ---------------------------------------------------------------------------
// Chain state — tracks the hash of the most recently INSERTED entry
// ---------------------------------------------------------------------------
let headHash = GENESIS_HASH;
let chainCounter = 0;

// ---------------------------------------------------------------------------
// Deterministic hash over the invariant fields of an entry
// ---------------------------------------------------------------------------
function computeChainHash(prevHash: string, entry: {
  id: string;
  timestamp: string;
  user: string;
  email: string;
  action: string;
  module: string;
  details: string;
  type: string;
}): string {
  const canonical = JSON.stringify({
    id: entry.id,
    timestamp: entry.timestamp,
    user: entry.user,
    email: entry.email,
    action: entry.action,
    module: entry.module,
    details: entry.details,
    type: entry.type,
  });
  return crypto.createHash('sha256').update(prevHash + canonical).digest('hex');
}

// ---------------------------------------------------------------------------
// Seed helpers — build chain-linked seed entries oldest→newest
// ---------------------------------------------------------------------------
function buildSeedEntry(
  base: Omit<AuditEntry, 'prevHash' | 'chainHash' | 'chainIndex'>,
  prevHashIn: string,
  index: number,
): AuditEntry {
  const chainHash = computeChainHash(prevHashIn, base);
  return { ...base, prevHash: prevHashIn, chainHash, chainIndex: index };
}

// Pre-seed with realistic demo entries — inserted oldest-first so chain is valid
const seed1Base = {
  id: 'seed-1',
  timestamp: new Date(Date.now() - 8 * 60 * 60 * 1000).toISOString(),
  user: 'Marcus Webb',
  email: 'm.webb@ligaturerd.io',
  action: 'User Login',
  module: 'Auth',
  details: 'Authenticated via SSO — session opened',
  type: 'login' as const,
};
const seed2Base = {
  id: 'seed-2',
  timestamp: new Date(Date.now() - 6 * 60 * 60 * 1000).toISOString(),
  user: 'Sarah Chen',
  email: 'sarah.chen@ligaturerd.io',
  action: 'Document Approved',
  module: 'Submissions',
  details: 'Approved CSR for LIG-2847 — three-stage review complete',
  type: 'document' as const,
};
const seed3Base = {
  id: 'seed-3',
  timestamp: new Date(Date.now() - 4 * 60 * 60 * 1000).toISOString(),
  user: 'James Liu',
  email: 'j.liu@ligaturerd.io',
  action: 'Status Changed',
  module: 'Safety',
  details: 'ICSR-2025-0088 status → Submitted (transmitted to EMA gateway)',
  type: 'document' as const,
};
const seed4Base = {
  id: 'seed-4',
  timestamp: new Date(Date.now() - 2 * 60 * 60 * 1000).toISOString(),
  user: 'Priya Anand',
  email: 'p.anand@ligaturerd.io',
  action: 'CAPA Created',
  module: 'QMS',
  details: 'CAPA-2026-0042 opened — deviation source: batch documentation gap',
  type: 'document' as const,
};

const s1 = buildSeedEntry(seed1Base, GENESIS_HASH, 0);
const s2 = buildSeedEntry(seed2Base, s1.chainHash, 1);
const s3 = buildSeedEntry(seed3Base, s2.chainHash, 2);
const s4 = buildSeedEntry(seed4Base, s3.chainHash, 3);

// Store is maintained newest-first for display; chain is built oldest-first
const auditStore: AuditEntry[] = [s4, s3, s2, s1];
headHash = s4.chainHash;
chainCounter = 4;

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function addAuditEntry(entry: Omit<AuditEntry, 'id' | 'timestamp' | 'prevHash' | 'chainHash' | 'chainIndex'>): AuditEntry {
  const id = 'evt-' + Date.now() + '-' + Math.random().toString(36).slice(2, 6);
  const timestamp = new Date().toISOString();
  const index = chainCounter;
  const chainHash = computeChainHash(headHash, { id, timestamp, ...entry });

  const newEntry: AuditEntry = {
    ...entry,
    id,
    timestamp,
    prevHash: headHash,
    chainHash,
    chainIndex: index,
  };

  headHash = chainHash;
  chainCounter += 1;

  auditStore.unshift(newEntry); // newest first for display
  if (auditStore.length > MAX_ENTRIES) auditStore.splice(MAX_ENTRIES);
  return newEntry;
}

export function getAuditEntries(limit = 50): AuditEntry[] {
  return auditStore.slice(0, limit);
}

export function getAuditCount(): number {
  return auditStore.length;
}

/**
 * Walk every stored entry oldest→newest, recompute each chainHash, verify
 * it matches the stored value AND that prevHash matches predecessor.
 */
export function verifyAuditChain(): ChainVerifyResult {
  const ordered = [...auditStore].sort((a, b) => a.chainIndex - b.chainIndex);
  let runningHash = GENESIS_HASH;

  for (const entry of ordered) {
    // Check prevHash linkage
    if (entry.prevHash !== runningHash) {
      return {
        valid: false,
        totalEntries: auditStore.length,
        verifiedAt: new Date().toISOString(),
        breakAt: entry.id,
        breakIndex: entry.chainIndex,
      };
    }
    // Recompute this entry's chainHash
    const expected = computeChainHash(runningHash, entry);
    if (entry.chainHash !== expected) {
      return {
        valid: false,
        totalEntries: auditStore.length,
        verifiedAt: new Date().toISOString(),
        breakAt: entry.id,
        breakIndex: entry.chainIndex,
      };
    }
    runningHash = entry.chainHash;
  }

  return {
    valid: true,
    totalEntries: auditStore.length,
    verifiedAt: new Date().toISOString(),
  };
}
